### ISSUE TITLE
Overhead with Views inside OpenMP loop

### ISSUE THREAD

**Issue by ArVictor** [kokkos/kokkos#1816]
I am trying to implement a parallel section using OpenMP for the loop and Views for the data structure. But I noticed that when, from my parallel loop, I call a function acting on a View:
```this->spline.view_1d(this->spline.view_coeffs)```
there are many calls to the Kokkos::Impl::SharedAllocationRecord::increment and decrement.
If I replace the function by the same function acting on a C++ array:
```this->spline.1d(this->spline.coeffs)```
those SharedAllocationRecord calls disappear and the performance is way better.

Would you have an explanation for that ? Are the Views considered "atomic" in my OpenMP loop ? Is there a solution ?

**Comment by crtrott** [comment]
Is this inside a `Kokkos::parallel_for` or a `#pragma omp parallel for`.

**Comment by ArVictor** [comment]
It is inside a ```#pragma omp parallel for```.
My goal is to implement one ```#pragma omp parallel for``` inside a code where I mainly use Kokkos' Views and loops.
So @stanmoore1's indication would mean duplicating all Views required in my loop as unmanaged.

**Comment by crtrott** [comment]
Yes. Inside a Kokkos::parallel_for we do turn reference counting off automatically. It happens during the dispatch phase of a functor but if you are using your own pragma omp parallel for this doesn't happen, which means your code is littered with atomic updates at every place where you assign views. 

As Stan said: use unmanaged views and you should be good.

**Comment by ArVictor** [comment]
Thank you, using unmanaged Views works perfectly!
But is it possible to turn reference counting off (and back on) manually? To know if I could avoid declaring new unmanaged views for all my existing views.

**Comment by mhoemmen** [comment]
Views are only made managed or unmanaged in their constructor, so as long as you don't construct new Views inside your parallel loop, you're fine.

**Comment by ArVictor** [comment]
I was asking to know if there was a more convenient way to have the "unmanaged behavior". But if managed/unmanaged is only set in the constructor, it answers my question :)

**Comment by crtrott** [comment]
Yeah there is no way to turn this on and off in different parts. While we could do that, it would get even further away of the spirit of C++ data containers etc.
